Compiler can't load an analyzer where a binding redirect may be needed

In https://github.com/microsoft/CsWin32/discussions/218, we had a report of a source generator that was failing to load in Omnisharp. Debugging revealed it the analyzer depends on System.Text.Json, and the NuGet package there depends on assembly version 4.1.3 of System.Numerics.Vectors. However, another dependency brings in 4.1.4 of System.Numerics.Vectors, which in a normal app would be resolved with a binding redirect at runtime. Since https://github.com/microsoft/CsWin32/issues/261 was fixed, the analyzer does package 4.1.4.

For our command line compilers, right now it works out that this analyzer will load, only because System.Numerics.Vectors 4.1.4 happens to be a dependency of the compiler and thus a direct is in place for it. But that's somewhat an implementation detail and could change. For Omnisharp however, it doesn't have a redirect and thus won't load. And there's no practical way to package both 4.1.3 and 4.1.4 versions of the dependency in the same NuGet package to work around the issues.

Discussing with @jaredpar it seems the best resolution here is to change our assembly loading rules a bit. When we load an analyzer, we require _exact matches_ of dependencies [here](https://github.com/dotnet/roslyn/blob/ac284e677662baaf9274eeb5dba45d2300ade0d6/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs#L188); changing that to allow an "or higher" would mean we'd be effectively doing a binding redirect at runtime, which would resolve these issues.

Omnisharp currently has a copy/paste of the analyzer assembly loading code in it's repo; once we have the compiler change in hand we should update that copy. Since we have an External Access shim, I'd propose we simply make an API for them to fetch the implementation Roslyn has and they can reuse that directly.

Another variant of the problem:

```
git clone https://github.com/dotnet/Nerdbank.GitVersioning.git
cd Nerdbank.GitVersioning
git checkout 69c33ed0be29de36e9fbd3188b55d0cd165d578a
.\init.ps1 -InstallLocality repo
dotnet build src\Nerdbank.GitVersioning
```

Fails with:

> CSC : warning CS8785: Generator 'SourceGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'FileLoadException' with message 'Could not load file or assembly 'System.Text.Json, Version=5.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'. Could not find or load a specific file. (0x80131621)' [D:\a\1\s\src\NerdBank.GitVersioning\NerdBank.GitVersioning.csproj]

For the System.Text.Json failure, it only repros when I have the .NETCoreApp3.1 runtime installed. Even though I'm using the 5.0 SDK, the compiler runs in a .NET Core 3.1 runtime process when it's available, and that's where it fails. Probably because assemblies that ship with the runtime won't be loaded from the app directory (or from the list of analyzer assemblies in my case) unless the .deps.json file tells the runtime to do so.
When I remove the .NET Core 3.1 runtime, the SDK compiler is forced to use the 5.0 runtime, and it works. You can try this yourself after reproing the problem by removing the `obj\tools\.dotnet\shared\Microsoft.NETCore.App\3.1.16` directory.

But I need the .NET Core 3.1 runtime for this repo/pipeline because I run tests on it.

--- a/src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs
+++ b/src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs
@@ -67,7 +67,7 @@ namespace Microsoft.CodeAnalysis.UnitTests
         {
             var loader = new DefaultAnalyzerAssemblyLoader();
 
-            Assert.Throws<ArgumentNullException>("fullPath", () => loader.AddDependencyLocation(null));
+            Assert.Throws<ArgumentNullException>("fullPath", () => loader.AddDependencyLocation(null!));
             Assert.Throws<ArgumentException>("fullPath", () => loader.AddDependencyLocation("a"));
         }
 

--- a/src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs
+++ b/src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs
@@ -326,6 +326,37 @@ Delta: Epsilon: Test E
             }
         }
 
+
+        [Fact]
+        public void AssemblyLoading_MultipleVersions_MultipleVersionsOfSameAnalyzerItself()
+        {
+            StringBuilder sb = new StringBuilder();
+
+            var loader = new DefaultAnalyzerAssemblyLoader();
+            loader.AddDependencyLocation(_testFixture.Delta2.Path);
+            loader.AddDependencyLocation(_testFixture.Delta2B.Path);
+
+            Assembly delta2 = loader.LoadFromPath(_testFixture.Delta2.Path);
+            Assembly delta2B = loader.LoadFromPath(_testFixture.Delta2B.Path);
+
+            // 2B or not 2B? That is the question...that depends on whether we're on .NET Core or not.
+
+#if NETCOREAPP
+
+            // On Core, we're able to load both of these into separate AssemblyLoadContexts.
+            Assert.NotEqual(delta2B.Location, delta2.Location);
+            Assert.Equal(_testFixture.Delta2.Path, delta2.Location);
+            Assert.Equal(_testFixture.Delta2B.Path, delta2B.Location);
+
+#else
+
+            // In non-core, we cache by assembly identity; since we don't use multiple AppDomains we have no
+            // way to load different assemblies with the same identity, no matter what. Thus, we'll get the
+            // same assembly for both of these.
+            Assert.Equal(delta2B, delta2);
+#endif
+        }
+
         [Fact]
         public void AssemblyLoading_MultipleVersions_ExactAndGreaterMatch()
         {

--- a/src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs
+++ b/src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs
@@ -353,7 +353,7 @@ Delta: Epsilon: Test E
             // In non-core, we cache by assembly identity; since we don't use multiple AppDomains we have no
             // way to load different assemblies with the same identity, no matter what. Thus, we'll get the
             // same assembly for both of these.
-            Assert.Equal(delta2B, delta2);
+            Assert.Same(delta2B, delta2);
 #endif
         }
 
@@ -408,6 +408,8 @@ Delta: Epsilon: Test E
             var delta1File = tempDir.CreateFile("Delta.dll").CopyContentFrom(_testFixture.Delta1.Path);
 
             // Epsilon wants Delta2, but since Delta1 is in the same directory, we prefer Delta1 over Delta2.
+            // This is because the CLR will see it first and load it, without giving us any chance to redirect
+            // in the AssemblyResolve hook.
             var loader = new DefaultAnalyzerAssemblyLoader();
             loader.AddDependencyLocation(delta1File.Path);
             loader.AddDependencyLocation(_testFixture.Delta2.Path);
@@ -514,6 +516,67 @@ Delta: Epsilon: Test E
             eWrite.Invoke(e, new object[] { sb, "Test E" });
             Assert.Equal(
 @"Delta: Gamma: Test G
+",
+                actual);
+        }
+
+        [Fact]
+        public void AssemblyLoading_UnifyToHighest()
+        {
+            var sb = new StringBuilder();
+
+            var loader = new DefaultAnalyzerAssemblyLoader();
+
+            // Gamma depends on Delta v1, and Epsilon depends on Delta v2. But both should load
+            // and both use Delta v2. We intentionally here are not adding a reference to Delta1, since
+            // this test is testing what happens if it's not present. A simple example for this scenario
+            // is an analyzer that depends on both Gamma and Epsilon; an analyzer package can't reasonably
+            // package both Delta v1 and Delta v2, so it'll only package the highest and things should work.
+            loader.AddDependencyLocation(_testFixture.GammaReferencingPublicSigned.Path);
+            loader.AddDependencyLocation(_testFixture.EpsilonReferencingPublicSigned.Path);
+            loader.AddDependencyLocation(_testFixture.DeltaPublicSigned2.Path);
+
+            var gamma = loader.LoadFromPath(_testFixture.GammaReferencingPublicSigned.Path);
+            var g = gamma.CreateInstance("Gamma.G")!;
+            g.GetType().GetMethod("Write")!.Invoke(g, new object[] { sb, "Test G" });
+
+            var epsilon = loader.LoadFromPath(_testFixture.EpsilonReferencingPublicSigned.Path);
+            var e = epsilon.CreateInstance("Epsilon.E")!;
+            e.GetType().GetMethod("Write")!.Invoke(e, new object[] { sb, "Test E" });
+
+            var actual = sb.ToString();
+
+            Assert.Equal(
+@"Delta.2: Gamma: Test G
+Delta.2: Epsilon: Test E
+",
+                actual);
+        }
+
+        [Fact]
+        public void AssemblyLoading_CanLoadDifferentVersionsDirectly()
+        {
+            var sb = new StringBuilder();
+
+            var loader = new DefaultAnalyzerAssemblyLoader();
+
+            // Ensure that no matter what, if we have two analyzers of different versions, we never unify them.
+            loader.AddDependencyLocation(_testFixture.DeltaPublicSigned1.Path);
+            loader.AddDependencyLocation(_testFixture.DeltaPublicSigned2.Path);
+
+            var delta1Assembly = loader.LoadFromPath(_testFixture.DeltaPublicSigned1.Path);
+            var delta1Instance = delta1Assembly.CreateInstance("Delta.D")!;
+            delta1Instance.GetType().GetMethod("Write")!.Invoke(delta1Instance, new object[] { sb, "Test D1" });
+
+            var delta2Assembly = loader.LoadFromPath(_testFixture.DeltaPublicSigned2.Path);
+            var delta2Instance = delta2Assembly.CreateInstance("Delta.D")!;
+            delta2Instance.GetType().GetMethod("Write")!.Invoke(delta2Instance, new object[] { sb, "Test D2" });
+
+            var actual = sb.ToString();
+
+            Assert.Equal(
+@"Delta: Test D1
+Delta.2: Test D2
 ",
                 actual);
         }

--- a/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs
+++ b/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs
@@ -2,12 +2,11 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-#nullable disable
-
 using System;
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
 using System.Reflection;
 using Roslyn.Utilities;
 
@@ -19,7 +18,7 @@ namespace Microsoft.CodeAnalysis
 
         // lock _guard to read/write
         private readonly Dictionary<string, Assembly> _loadedAssembliesByPath = new();
-        private readonly Dictionary<string, AssemblyIdentity> _loadedAssemblyIdentitiesByPath = new();
+        private readonly Dictionary<string, AssemblyIdentity?> _loadedAssemblyIdentitiesByPath = new();
         private readonly Dictionary<AssemblyIdentity, Assembly> _loadedAssembliesByIdentity = new();
 
         // maps file name to a full path (lock _guard to read/write):
@@ -59,12 +58,12 @@ namespace Microsoft.CodeAnalysis
             return LoadFromPathUncheckedCore(fullPath);
         }
 
-        private Assembly LoadFromPathUncheckedCore(string fullPath, AssemblyIdentity identity = null)
+        private Assembly LoadFromPathUncheckedCore(string fullPath, AssemblyIdentity? identity = null)
         {
             Debug.Assert(PathUtilities.IsAbsolute(fullPath));
 
             // Check if we have already loaded an assembly with the same identity or from the given path.
-            Assembly loadedAssembly = null;
+            Assembly? loadedAssembly = null;
             lock (_guard)
             {
                 if (_loadedAssembliesByPath.TryGetValue(fullPath, out var existingAssembly))
@@ -91,7 +90,7 @@ namespace Microsoft.CodeAnalysis
             return AddToCache(loadedAssembly, fullPath, identity);
         }
 
-        private Assembly AddToCache(Assembly assembly, string fullPath, AssemblyIdentity identity)
+        private Assembly AddToCache(Assembly assembly, string fullPath, AssemblyIdentity? identity)
         {
             Debug.Assert(PathUtilities.IsAbsolute(fullPath));
             Debug.Assert(assembly != null);
@@ -120,7 +119,7 @@ namespace Microsoft.CodeAnalysis
             }
         }
 
-        private AssemblyIdentity GetOrAddAssemblyIdentity(string fullPath)
+        private AssemblyIdentity? GetOrAddAssemblyIdentity(string fullPath)
         {
             Debug.Assert(PathUtilities.IsAbsolute(fullPath));
 
@@ -136,7 +135,8 @@ namespace Microsoft.CodeAnalysis
             return AddToCache(fullPath, identity);
         }
 
-        private AssemblyIdentity AddToCache(string fullPath, AssemblyIdentity identity)
+        [return: NotNullIfNotNull("identity")]
+        private AssemblyIdentity? AddToCache(string fullPath, AssemblyIdentity? identity)
         {
             lock (_guard)
             {
@@ -153,7 +153,6 @@ namespace Microsoft.CodeAnalysis
             return identity;
         }
 
-#nullable enable
         protected HashSet<string>? GetPaths(string simpleName)
         {
             _knownAssemblyPathsBySimpleName.TryGetValue(simpleName, out var paths);
@@ -169,9 +168,8 @@ namespace Microsoft.CodeAnalysis
         {
             return fullPath;
         }
-#nullable disable
 
-        public Assembly Load(string displayName)
+        public Assembly? Load(string displayName)
         {
             if (!AssemblyIdentity.TryParseDisplayName(displayName, out var requestedIdentity))
             {

--- a/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs
+++ b/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs
@@ -6,7 +6,6 @@ using System;
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Diagnostics;
-using System.Diagnostics.CodeAnalysis;
 using System.Reflection;
 using Roslyn.Utilities;
 
@@ -20,7 +19,7 @@ namespace Microsoft.CodeAnalysis
         private readonly Dictionary<string, Assembly> _loadedAssembliesByPath = new();
 
         // maps file name to a full path (lock _guard to read/write):
-        private readonly Dictionary<string, HashSet<string>> _knownAssemblyPathsBySimpleName = new(StringComparer.OrdinalIgnoreCase);
+        private readonly Dictionary<string, ImmutableHashSet<string>> _knownAssemblyPathsBySimpleName = new(StringComparer.OrdinalIgnoreCase);
 
         protected abstract Assembly LoadFromPathImpl(string fullPath);
 
@@ -35,11 +34,13 @@ namespace Microsoft.CodeAnalysis
             {
                 if (!_knownAssemblyPathsBySimpleName.TryGetValue(simpleName, out var paths))
                 {
-                    paths = new HashSet<string>(PathUtilities.Comparer);
+                    paths = ImmutableHashSet.Create(PathUtilities.Comparer, fullPath);
                     _knownAssemblyPathsBySimpleName.Add(simpleName, paths);
                 }
-
-                paths.Add(fullPath);
+                else
+                {
+                    _knownAssemblyPathsBySimpleName[simpleName] = paths.Add(fullPath);
+                }
             }
         }
 
@@ -80,10 +81,13 @@ namespace Microsoft.CodeAnalysis
             return loadedAssembly;
         }
 
-        protected HashSet<string>? GetPaths(string simpleName)
+        protected ImmutableHashSet<string>? GetPaths(string simpleName)
         {
-            _knownAssemblyPathsBySimpleName.TryGetValue(simpleName, out var paths);
-            return paths;
+            lock (_guard)
+            {
+                _knownAssemblyPathsBySimpleName.TryGetValue(simpleName, out var paths);
+                return paths;
+            }
         }
 
         /// <summary>

--- a/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs
+++ b/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs
@@ -11,6 +11,14 @@ using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis
 {
+    /// <summary>
+    /// The base implementation for <see cref="IAnalyzerAssemblyLoader"/>. This type provides caching and tracking of inputs given
+    /// to <see cref="AddDependencyLocation(string)"/>.
+    /// </summary>
+    /// <remarks>
+    /// This type generally assumes that files on disk aren't changing, since it ensure that two calls to <see cref="LoadFromPath(string)"/>
+    /// will always return the same thing, per that interface's contract.
+    /// </remarks>
     internal abstract class AnalyzerAssemblyLoader : IAnalyzerAssemblyLoader
     {
         private readonly object _guard = new();
@@ -21,7 +29,10 @@ namespace Microsoft.CodeAnalysis
         // maps file name to a full path (lock _guard to read/write):
         private readonly Dictionary<string, ImmutableHashSet<string>> _knownAssemblyPathsBySimpleName = new(StringComparer.OrdinalIgnoreCase);
 
-        protected abstract Assembly LoadFromPathImpl(string fullPath);
+        /// <summary>
+        /// Implemented by derived types to actually perform the load for an assembly that doesn't have a cached result.
+        /// </summary>
+        protected abstract Assembly LoadFromPathUncheckedImpl(string fullPath);
 
         #region Public API
 
@@ -52,6 +63,10 @@ namespace Microsoft.CodeAnalysis
 
         #endregion
 
+        /// <summary>
+        /// Returns the cached assembly for fullPath if we've done a load for this path before, or calls <see cref="LoadFromPathUncheckedImpl"/> if
+        /// it needs to be loaded. This method skips the check in release builds that the path is an absolute path, hence the "Unchecked" in the name.
+        /// </summary>
         protected Assembly LoadFromPathUnchecked(string fullPath)
         {
             Debug.Assert(PathUtilities.IsAbsolute(fullPath));
@@ -69,7 +84,7 @@ namespace Microsoft.CodeAnalysis
             // Otherwise, load the assembly.
             if (loadedAssembly == null)
             {
-                loadedAssembly = LoadFromPathImpl(fullPath);
+                loadedAssembly = LoadFromPathUncheckedImpl(fullPath);
             }
 
             // Add the loaded assembly to the path cache.

--- a/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs
+++ b/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs
@@ -18,8 +18,6 @@ namespace Microsoft.CodeAnalysis
 
         // lock _guard to read/write
         private readonly Dictionary<string, Assembly> _loadedAssembliesByPath = new();
-        private readonly Dictionary<string, AssemblyIdentity?> _loadedAssemblyIdentitiesByPath = new();
-        private readonly Dictionary<AssemblyIdentity, Assembly> _loadedAssembliesByIdentity = new();
 
         // maps file name to a full path (lock _guard to read/write):
         private readonly Dictionary<string, HashSet<string>> _knownAssemblyPathsBySimpleName = new(StringComparer.OrdinalIgnoreCase);
@@ -53,25 +51,17 @@ namespace Microsoft.CodeAnalysis
 
         #endregion
 
-        private Assembly LoadFromPathUnchecked(string fullPath, AssemblyIdentity? identity = null)
+        protected Assembly LoadFromPathUnchecked(string fullPath)
         {
             Debug.Assert(PathUtilities.IsAbsolute(fullPath));
 
-            // Check if we have already loaded an assembly with the same identity or from the given path.
+            // Check if we have already loaded an assembly from the given path.
             Assembly? loadedAssembly = null;
             lock (_guard)
             {
                 if (_loadedAssembliesByPath.TryGetValue(fullPath, out var existingAssembly))
                 {
                     loadedAssembly = existingAssembly;
-                }
-                else
-                {
-                    identity ??= GetOrAddAssemblyIdentity(fullPath);
-                    if (identity != null && _loadedAssembliesByIdentity.TryGetValue(identity, out existingAssembly))
-                    {
-                        loadedAssembly = existingAssembly;
-                    }
                 }
             }
 
@@ -81,71 +71,13 @@ namespace Microsoft.CodeAnalysis
                 loadedAssembly = LoadFromPathImpl(fullPath);
             }
 
-            // Add the loaded assembly to both path and identity cache.
-            return AddToCache(loadedAssembly, fullPath, identity);
-        }
-
-        private Assembly AddToCache(Assembly assembly, string fullPath, AssemblyIdentity? identity)
-        {
-            Debug.Assert(PathUtilities.IsAbsolute(fullPath));
-            Debug.Assert(assembly != null);
-
-            identity = AddToCache(fullPath, identity ?? AssemblyIdentity.FromAssemblyDefinition(assembly));
-            Debug.Assert(identity != null);
-
+            // Add the loaded assembly to the path cache.
             lock (_guard)
             {
-                // The same assembly may be loaded from two different full paths (e.g. when loaded from GAC, etc.),
-                // or another thread might have loaded the assembly after we checked above.
-                if (_loadedAssembliesByIdentity.TryGetValue(identity, out var existingAssembly))
-                {
-                    assembly = existingAssembly;
-                }
-                else
-                {
-                    _loadedAssembliesByIdentity.Add(identity, assembly);
-                }
-
-                // An assembly file might be replaced by another file with a different identity.
-                // Last one wins.
-                _loadedAssembliesByPath[fullPath] = assembly;
-
-                return assembly;
-            }
-        }
-
-        private AssemblyIdentity? GetOrAddAssemblyIdentity(string fullPath)
-        {
-            Debug.Assert(PathUtilities.IsAbsolute(fullPath));
-
-            lock (_guard)
-            {
-                if (_loadedAssemblyIdentitiesByPath.TryGetValue(fullPath, out var existingIdentity))
-                {
-                    return existingIdentity;
-                }
+                _loadedAssembliesByPath[fullPath] = loadedAssembly;
             }
 
-            var identity = AssemblyIdentityUtils.TryGetAssemblyIdentity(fullPath);
-            return AddToCache(fullPath, identity);
-        }
-
-        [return: NotNullIfNotNull("identity")]
-        private AssemblyIdentity? AddToCache(string fullPath, AssemblyIdentity? identity)
-        {
-            lock (_guard)
-            {
-                if (_loadedAssemblyIdentitiesByPath.TryGetValue(fullPath, out var existingIdentity) && existingIdentity != null)
-                {
-                    identity = existingIdentity;
-                }
-                else
-                {
-                    _loadedAssemblyIdentitiesByPath[fullPath] = identity;
-                }
-            }
-
-            return identity;
+            return loadedAssembly;
         }
 
         protected HashSet<string>? GetPaths(string simpleName)
@@ -162,47 +94,6 @@ namespace Microsoft.CodeAnalysis
         protected virtual string GetPathToLoad(string fullPath)
         {
             return fullPath;
-        }
-
-        public Assembly? Load(string displayName)
-        {
-            if (!AssemblyIdentity.TryParseDisplayName(displayName, out var requestedIdentity))
-            {
-                return null;
-            }
-
-            ImmutableArray<string> candidatePaths;
-            lock (_guard)
-            {
-
-                // First, check if this loader already loaded the requested assembly:
-                if (_loadedAssembliesByIdentity.TryGetValue(requestedIdentity, out var existingAssembly))
-                {
-                    return existingAssembly;
-                }
-                // Second, check if an assembly file of the same simple name was registered with the loader:
-                if (!_knownAssemblyPathsBySimpleName.TryGetValue(requestedIdentity.Name, out var pathList))
-                {
-                    return null;
-                }
-
-                Debug.Assert(pathList.Count > 0);
-                candidatePaths = pathList.ToImmutableArray();
-            }
-
-            // Multiple assemblies of the same simple name but different identities might have been registered.
-            // Load the one that matches the requested identity (if any).
-            foreach (var candidatePath in candidatePaths)
-            {
-                var candidateIdentity = GetOrAddAssemblyIdentity(candidatePath);
-
-                if (requestedIdentity.Equals(candidateIdentity))
-                {
-                    return LoadFromPathUnchecked(candidatePath, candidateIdentity);
-                }
-            }
-
-            return null;
         }
     }
 }

--- a/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs
+++ b/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs
@@ -53,12 +53,7 @@ namespace Microsoft.CodeAnalysis
 
         #endregion
 
-        private Assembly LoadFromPathUnchecked(string fullPath)
-        {
-            return LoadFromPathUncheckedCore(fullPath);
-        }
-
-        private Assembly LoadFromPathUncheckedCore(string fullPath, AssemblyIdentity? identity = null)
+        private Assembly LoadFromPathUnchecked(string fullPath, AssemblyIdentity? identity = null)
         {
             Debug.Assert(PathUtilities.IsAbsolute(fullPath));
 
@@ -203,7 +198,7 @@ namespace Microsoft.CodeAnalysis
 
                 if (requestedIdentity.Equals(candidateIdentity))
                 {
-                    return LoadFromPathUncheckedCore(candidatePath, candidateIdentity);
+                    return LoadFromPathUnchecked(candidatePath, candidateIdentity);
                 }
             }
 

--- a/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs
+++ b/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs
@@ -162,7 +162,7 @@ namespace Microsoft.CodeAnalysis
         /// <summary>
         /// When overridden in a derived class, allows substituting an assembly path after we've
         /// identified the context to load an assembly in, but before the assembly is actually
-        /// loaded from disk.
+        /// loaded from disk. This is used to substitute out the original path with the shadow-copied version.
         /// </summary>
         protected virtual string GetPathToLoad(string fullPath)
         {

--- a/src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs
+++ b/src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs
@@ -79,7 +79,7 @@ namespace Microsoft.CodeAnalysis
         private readonly object _guard = new object();
         private readonly Dictionary<string, DirectoryLoadContext> _loadContextByDirectory = new Dictionary<string, DirectoryLoadContext>(StringComparer.Ordinal);
 
-        protected override Assembly LoadFromPathImpl(string fullPath)
+        protected override Assembly LoadFromPathUncheckedImpl(string fullPath)
         {
             DirectoryLoadContext? loadContext;
 

--- a/src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Desktop.cs
+++ b/src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Desktop.cs
@@ -2,8 +2,6 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-#nullable disable
-
 #if !NETCOREAPP
 
 using System;
@@ -38,7 +36,7 @@ namespace Microsoft.CodeAnalysis
             return Assembly.LoadFrom(pathToLoad);
         }
 
-        private Assembly CurrentDomain_AssemblyResolve(object sender, ResolveEventArgs args)
+        private Assembly? CurrentDomain_AssemblyResolve(object sender, ResolveEventArgs args)
         {
             try
             {

--- a/src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Desktop.cs
+++ b/src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Desktop.cs
@@ -5,8 +5,13 @@
 #if !NETCOREAPP
 
 using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
 using System.Reflection;
 using System.Threading;
+using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis
 {
@@ -23,6 +28,11 @@ namespace Microsoft.CodeAnalysis
     /// </remarks>
     internal class DefaultAnalyzerAssemblyLoader : AnalyzerAssemblyLoader
     {
+        private readonly object _guard = new();
+
+        private readonly Dictionary<AssemblyIdentity, Assembly> _loadedAssembliesByIdentity = new();
+        private readonly Dictionary<string, AssemblyIdentity?> _loadedAssemblyIdentitiesByPath = new();
+
         private int _hookedAssemblyResolve;
 
         protected override Assembly LoadFromPathImpl(string fullPath)
@@ -32,8 +42,37 @@ namespace Microsoft.CodeAnalysis
                 AppDomain.CurrentDomain.AssemblyResolve += CurrentDomain_AssemblyResolve;
             }
 
+            AssemblyIdentity? identity;
+
+            lock (_guard)
+            {
+                identity = GetOrAddAssemblyIdentity(fullPath);
+                if (identity != null && _loadedAssembliesByIdentity.TryGetValue(identity, out var existingAssembly))
+                {
+                    return existingAssembly;
+                }
+            }
+
             var pathToLoad = GetPathToLoad(fullPath);
-            return Assembly.LoadFrom(pathToLoad);
+            var loadedAssembly = Assembly.LoadFrom(pathToLoad);
+
+            lock (_guard)
+            {
+                identity ??= identity ?? AssemblyIdentity.FromAssemblyDefinition(loadedAssembly);
+
+                // The same assembly may be loaded from two different full paths (e.g. when loaded from GAC, etc.),
+                // or another thread might have loaded the assembly after we checked above.
+                if (_loadedAssembliesByIdentity.TryGetValue(identity, out var existingAssembly))
+                {
+                    loadedAssembly = existingAssembly;
+                }
+                else
+                {
+                    _loadedAssembliesByIdentity.Add(identity, loadedAssembly);
+                }
+
+                return loadedAssembly;
+            }
         }
 
         private Assembly? CurrentDomain_AssemblyResolve(object sender, ResolveEventArgs args)
@@ -46,6 +85,82 @@ namespace Microsoft.CodeAnalysis
             {
                 return null;
             }
+        }
+
+        private AssemblyIdentity? GetOrAddAssemblyIdentity(string fullPath)
+        {
+            Debug.Assert(PathUtilities.IsAbsolute(fullPath));
+
+            lock (_guard)
+            {
+                if (_loadedAssemblyIdentitiesByPath.TryGetValue(fullPath, out var existingIdentity))
+                {
+                    return existingIdentity;
+                }
+            }
+
+            var identity = AssemblyIdentityUtils.TryGetAssemblyIdentity(fullPath);
+            return AddToCache(fullPath, identity);
+        }
+
+        [return: NotNullIfNotNull("identity")]
+        private AssemblyIdentity? AddToCache(string fullPath, AssemblyIdentity? identity)
+        {
+            lock (_guard)
+            {
+                if (_loadedAssemblyIdentitiesByPath.TryGetValue(fullPath, out var existingIdentity) && existingIdentity != null)
+                {
+                    identity = existingIdentity;
+                }
+                else
+                {
+                    _loadedAssemblyIdentitiesByPath[fullPath] = identity;
+                }
+            }
+
+            return identity;
+        }
+
+        private Assembly? Load(string displayName)
+        {
+            if (!AssemblyIdentity.TryParseDisplayName(displayName, out var requestedIdentity))
+            {
+                return null;
+            }
+
+            ImmutableArray<string> candidatePaths;
+            lock (_guard)
+            {
+
+                // First, check if this loader already loaded the requested assembly:
+                if (_loadedAssembliesByIdentity.TryGetValue(requestedIdentity, out var existingAssembly))
+                {
+                    return existingAssembly;
+                }
+                // Second, check if an assembly file of the same simple name was registered with the loader:
+                var pathList = GetPaths(requestedIdentity.Name);
+                if (pathList is null)
+                {
+                    return null;
+                }
+
+                Debug.Assert(pathList.Count > 0);
+                candidatePaths = pathList.ToImmutableArray();
+            }
+
+            // Multiple assemblies of the same simple name but different identities might have been registered.
+            // Load the one that matches the requested identity (if any).
+            foreach (var candidatePath in candidatePaths)
+            {
+                var candidateIdentity = GetOrAddAssemblyIdentity(candidatePath);
+
+                if (requestedIdentity.Equals(candidateIdentity))
+                {
+                    return LoadFromPathUnchecked(candidatePath);
+                }
+            }
+
+            return null;
         }
     }
 }

--- a/src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Desktop.cs
+++ b/src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Desktop.cs
@@ -149,7 +149,9 @@ namespace Microsoft.CodeAnalysis
             string? bestPath = null;
             Version? bestIdentityVersion = null;
 
-            foreach (var candidatePath in candidatePaths)
+            // Sort the candidate paths by ordinal, to ensure determinism with the same inputs if you were to have multiple assemblies
+            // providing the same version.
+            foreach (var candidatePath in candidatePaths.OrderBy(StringComparer.Ordinal))
             {
                 var candidateIdentity = GetOrAddAssemblyIdentity(candidatePath);
 

--- a/src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Desktop.cs
+++ b/src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Desktop.cs
@@ -101,16 +101,12 @@ namespace Microsoft.CodeAnalysis
             }
 
             var identity = AssemblyIdentityUtils.TryGetAssemblyIdentity(fullPath);
-            return AddToCache(fullPath, identity);
-        }
 
-        [return: NotNullIfNotNull("identity")]
-        private AssemblyIdentity? AddToCache(string fullPath, AssemblyIdentity? identity)
-        {
             lock (_guard)
             {
                 if (_loadedAssemblyIdentitiesByPath.TryGetValue(fullPath, out var existingIdentity) && existingIdentity != null)
                 {
+                    // Somebody else beat us, so used the cached value
                     identity = existingIdentity;
                 }
                 else

--- a/src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Desktop.cs
+++ b/src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Desktop.cs
@@ -9,6 +9,7 @@ using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Diagnostics;
 using System.Diagnostics.CodeAnalysis;
+using System.Linq;
 using System.Reflection;
 using System.Threading;
 using Roslyn.Utilities;
@@ -147,16 +148,30 @@ namespace Microsoft.CodeAnalysis
                 Debug.Assert(candidatePaths.Count > 0);
             }
 
-            // Multiple assemblies of the same simple name but different identities might have been registered.
-            // Load the one that matches the requested identity (if any).
+            // Find the highest version that satisfies the original request. We'll match for the highest version we can, assuming it
+            // actually matches the original request
+            string? bestPath = null;
+            Version? bestIdentityVersion = null;
+
             foreach (var candidatePath in candidatePaths)
             {
                 var candidateIdentity = GetOrAddAssemblyIdentity(candidatePath);
 
-                if (requestedIdentity.Equals(candidateIdentity))
+                if (candidateIdentity is not null &&
+                    candidateIdentity.Version >= requestedIdentity.Version &&
+                    candidateIdentity.PublicKeyToken.SequenceEqual(requestedIdentity.PublicKeyToken))
                 {
-                    return LoadFromPathUnchecked(candidatePath);
+                    if (bestIdentityVersion is null || candidateIdentity.Version > bestIdentityVersion)
+                    {
+                        bestPath = candidatePath;
+                        bestIdentityVersion = candidateIdentity.Version;
+                    }
                 }
+            }
+
+            if (bestPath != null)
+            {
+                return LoadFromPathUnchecked(bestPath);
             }
 
             return null;

--- a/src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Desktop.cs
+++ b/src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Desktop.cs
@@ -128,7 +128,7 @@ namespace Microsoft.CodeAnalysis
                 return null;
             }
 
-            ImmutableArray<string> candidatePaths;
+            ImmutableHashSet<string> candidatePaths;
             lock (_guard)
             {
 
@@ -138,14 +138,13 @@ namespace Microsoft.CodeAnalysis
                     return existingAssembly;
                 }
                 // Second, check if an assembly file of the same simple name was registered with the loader:
-                var pathList = GetPaths(requestedIdentity.Name);
-                if (pathList is null)
+                candidatePaths = GetPaths(requestedIdentity.Name);
+                if (candidatePaths is null)
                 {
                     return null;
                 }
 
-                Debug.Assert(pathList.Count > 0);
-                candidatePaths = pathList.ToImmutableArray();
+                Debug.Assert(candidatePaths.Count > 0);
             }
 
             // Multiple assemblies of the same simple name but different identities might have been registered.

--- a/src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Desktop.cs
+++ b/src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Desktop.cs
@@ -80,7 +80,7 @@ namespace Microsoft.CodeAnalysis
         {
             try
             {
-                return Load(AppDomain.CurrentDomain.ApplyPolicy(args.Name));
+                return GetOrLoad(AppDomain.CurrentDomain.ApplyPolicy(args.Name));
             }
             catch
             {
@@ -122,7 +122,7 @@ namespace Microsoft.CodeAnalysis
             return identity;
         }
 
-        private Assembly? Load(string displayName)
+        private Assembly? GetOrLoad(string displayName)
         {
             if (!AssemblyIdentity.TryParseDisplayName(displayName, out var requestedIdentity))
             {

--- a/src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Desktop.cs
+++ b/src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Desktop.cs
@@ -36,7 +36,7 @@ namespace Microsoft.CodeAnalysis
 
         private int _hookedAssemblyResolve;
 
-        protected override Assembly LoadFromPathImpl(string fullPath)
+        protected override Assembly LoadFromPathUncheckedImpl(string fullPath)
         {
             if (Interlocked.CompareExchange(ref _hookedAssemblyResolve, 0, 1) == 0)
             {

--- a/src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Desktop.cs
+++ b/src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Desktop.cs
@@ -78,6 +78,9 @@ namespace Microsoft.CodeAnalysis
 
         private Assembly? CurrentDomain_AssemblyResolve(object sender, ResolveEventArgs args)
         {
+            // In the .NET Framework, if a handler to AssemblyResolve throws an exception, other handlers
+            // are not called. To avoid any bug in our handler breaking other handlers running in the same process
+            // we catch exceptions here. We do not expect exceptions to be thrown though.
             try
             {
                 return GetOrLoad(AppDomain.CurrentDomain.ApplyPolicy(args.Name));

--- a/src/Compilers/Core/Portable/DiagnosticAnalyzer/ShadowCopyAnalyzerAssemblyLoader.cs
+++ b/src/Compilers/Core/Portable/DiagnosticAnalyzer/ShadowCopyAnalyzerAssemblyLoader.cs
@@ -2,12 +2,9 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-#nullable disable
-
 using System;
 using System.Collections.Generic;
 using System.IO;
-using System.Reflection;
 using System.Threading;
 using System.Threading.Tasks;
 
@@ -36,7 +33,7 @@ namespace Microsoft.CodeAnalysis
         /// </summary>
         private int _assemblyDirectoryId;
 
-        public ShadowCopyAnalyzerAssemblyLoader(string baseDirectory = null)
+        public ShadowCopyAnalyzerAssemblyLoader(string? baseDirectory = null)
         {
             if (baseDirectory != null)
             {
@@ -72,7 +69,7 @@ namespace Microsoft.CodeAnalysis
             foreach (var subDirectory in subDirectories)
             {
                 string name = Path.GetFileName(subDirectory).ToLowerInvariant();
-                Mutex mutex = null;
+                Mutex? mutex = null;
                 try
                 {
                     // We only want to try deleting the directory if no-one else is currently
@@ -98,14 +95,12 @@ namespace Microsoft.CodeAnalysis
             }
         }
 
-#nullable enable
         protected override string GetPathToLoad(string fullPath)
         {
             string assemblyDirectory = CreateUniqueDirectoryForAssembly();
             string shadowCopyPath = CopyFileAndResources(fullPath, assemblyDirectory);
             return shadowCopyPath;
         }
-#nullable disable
 
         private static string CopyFileAndResources(string fullPath, string assemblyDirectory)
         {
@@ -114,7 +109,7 @@ namespace Microsoft.CodeAnalysis
 
             CopyFile(fullPath, shadowCopyPath);
 
-            string originalDirectory = Path.GetDirectoryName(fullPath);
+            string originalDirectory = Path.GetDirectoryName(fullPath)!;
             string fileNameWithoutExtension = Path.GetFileNameWithoutExtension(fileNameWithExtension);
             string resourcesNameWithoutExtension = fileNameWithoutExtension + ".resources";
             string resourcesNameWithExtension = resourcesNameWithoutExtension + ".dll";

--- a/src/Compilers/Test/Core/AssemblyLoadTestFixture.cs
+++ b/src/Compilers/Test/Core/AssemblyLoadTestFixture.cs
@@ -15,8 +15,6 @@ namespace Microsoft.CodeAnalysis.Test.Utilities
 {
     public sealed class AssemblyLoadTestFixture : IDisposable
     {
-        private static readonly CSharpCompilationOptions s_dllWithMaxWarningLevel = new(OutputKind.DynamicallyLinkedLibrary, warningLevel: Diagnostic.MaxWarningLevel);
-
         private readonly TempRoot _temp;
         private readonly TempDirectory _directory;
 
@@ -26,9 +24,19 @@ namespace Microsoft.CodeAnalysis.Test.Utilities
         public TempFile Delta1 { get; }
 
         /// <summary>
+        /// An assembly with no references, assembly version 1, and public signed.
+        /// </summary>
+        public TempFile DeltaPublicSigned1 { get; }
+
+        /// <summary>
         /// An assembly with a reference to <see cref="Delta1"/>.
         /// </summary>
         public TempFile Gamma { get; }
+
+        /// <summary>
+        /// An assembly with a reference to <see cref="DeltaPublicSigned1"/>.
+        /// </summary>
+        public TempFile GammaReferencingPublicSigned { get; }
 
         /// <summary>
         /// An assembly with a reference to <see cref="Gamma"/>.
@@ -46,9 +54,19 @@ namespace Microsoft.CodeAnalysis.Test.Utilities
         public TempFile Delta2 { get; }
 
         /// <summary>
+        /// An assembly with no references, assembly version 2, and public signed.
+        /// </summary>
+        public TempFile DeltaPublicSigned2 { get; }
+
+        /// <summary>
         /// An assembly with a reference to <see cref="Delta2"/>.
         /// </summary>
         public TempFile Epsilon { get; }
+
+        /// <summary>
+        /// An assembly with a reference to <see cref="DeltaPublicSigned2"/>.
+        /// </summary>
+        public TempFile EpsilonReferencingPublicSigned { get; }
 
         /// <summary>
         /// An assembly with no references, assembly version 2. The implementation however is different than
@@ -89,7 +107,7 @@ namespace Microsoft.CodeAnalysis.Test.Utilities
             _temp = new TempRoot();
             _directory = _temp.CreateDirectory();
 
-            Delta1 = GenerateDll("Delta", _directory, @"
+            const string Delta1Source = @"
 using System.Text;
 
 [assembly: System.Reflection.AssemblyTitle(""Delta"")]
@@ -105,9 +123,13 @@ namespace Delta
         }
     }
 }
-");
+";
+
+            Delta1 = GenerateDll("Delta", _directory, Delta1Source);
             var delta1Reference = MetadataReference.CreateFromFile(Delta1.Path);
-            Gamma = GenerateDll("Gamma", _directory, @"
+            DeltaPublicSigned1 = GenerateDll("DeltaPublicSigned", _directory.CreateDirectory("Delta1PublicSigned"), Delta1Source, publicSign: true);
+
+            const string GammaSource = @"
 using System.Text;
 using Delta;
 
@@ -123,7 +145,9 @@ namespace Gamma
         }
     }
 }
-", delta1Reference);
+";
+            Gamma = GenerateDll("Gamma", _directory, GammaSource, delta1Reference);
+            GammaReferencingPublicSigned = GenerateDll("GammaReferencingPublicSigned", _directory.CreateDirectory("GammaReferencingPublicSigned"), GammaSource, MetadataReference.CreateFromFile(DeltaPublicSigned1.Path));
 
             var gammaReference = MetadataReference.CreateFromFile(Gamma.Path);
             Beta = GenerateDll("Beta", _directory, @"
@@ -161,8 +185,7 @@ namespace Alpha
 }
 ", gammaReference);
 
-            var v2Directory = _directory.CreateDirectory("Version2");
-            Delta2 = GenerateDll("Delta", v2Directory, @"
+            const string Delta2Source = @"
 using System.Text;
 
 [assembly: System.Reflection.AssemblyTitle(""Delta"")]
@@ -178,9 +201,16 @@ namespace Delta
         }
     }
 }
-");
+";
+
+            var v2Directory = _directory.CreateDirectory("Version2");
+            Delta2 = GenerateDll("Delta", v2Directory, Delta2Source);
+            var v2PublicSignedDirectory = _directory.CreateDirectory("Version2PublicSigned");
+            DeltaPublicSigned2 = GenerateDll("DeltaPublicSigned", v2PublicSignedDirectory, Delta2Source, publicSign: true);
+
             var delta2Reference = MetadataReference.CreateFromFile(Delta2.Path);
-            Epsilon = GenerateDll("Epsilon", v2Directory, @"
+
+            const string EpsilonSource = @"
 using System.Text;
 using Delta;
 
@@ -196,7 +226,9 @@ namespace Epsilon
         }
     }
 }
-", delta2Reference);
+";
+            Epsilon = GenerateDll("Epsilon", v2Directory, EpsilonSource, delta2Reference);
+            EpsilonReferencingPublicSigned = GenerateDll("EpsilonReferencingPublicSigned", v2PublicSignedDirectory, EpsilonSource, MetadataReference.CreateFromFile(DeltaPublicSigned2.Path));
 
             var v2BDirectory = _directory.CreateDirectory("Version2B");
             Delta2B = GenerateDll("Delta", v2BDirectory, @"
@@ -357,16 +389,28 @@ public class Class1
 
         private static TempFile GenerateDll(string assemblyName, TempDirectory directory, string csSource, params MetadataReference[] additionalReferences)
         {
+            return GenerateDll(assemblyName, directory, csSource, publicSign: false, additionalReferences);
+        }
+
+        private static TempFile GenerateDll(string assemblyName, TempDirectory directory, string csSource, bool publicSign, params MetadataReference[] additionalReferences)
+        {
+            CSharpCompilationOptions options = new(OutputKind.DynamicallyLinkedLibrary, warningLevel: Diagnostic.MaxWarningLevel);
+
+            if (publicSign)
+            {
+                options = options.WithPublicSign(true).WithCryptoPublicKey(SigningTestHelpers.PublicKey);
+            }
+
             var analyzerDependencyCompilation = CSharpCompilation.Create(
                 assemblyName: assemblyName,
                 syntaxTrees: new SyntaxTree[] { SyntaxFactory.ParseSyntaxTree(csSource) },
-                references: new MetadataReference[]
+                references: (new MetadataReference[]
                 {
                     NetStandard20.mscorlib,
                     NetStandard20.netstandard,
                     NetStandard20.SystemRuntime
-                }.Concat(additionalReferences),
-                options: s_dllWithMaxWarningLevel);
+                }).Concat(additionalReferences),
+                options: options);
 
             var tempFile = directory.CreateFile($"{assemblyName}.dll");
             tempFile.WriteAllBytes(analyzerDependencyCompilation.EmitToArray());

--- a/src/Compilers/Test/Core/AssemblyLoadTestFixture.cs
+++ b/src/Compilers/Test/Core/AssemblyLoadTestFixture.cs
@@ -20,15 +20,45 @@ namespace Microsoft.CodeAnalysis.Test.Utilities
         private readonly TempRoot _temp;
         private readonly TempDirectory _directory;
 
+        /// <summary>
+        /// An assembly with no references, assembly version 1.
+        /// </summary>
         public TempFile Delta1 { get; }
+
+        /// <summary>
+        /// An assembly with a reference to <see cref="Delta1"/>.
+        /// </summary>
         public TempFile Gamma { get; }
+
+        /// <summary>
+        /// An assembly with a reference to <see cref="Gamma"/>.
+        /// </summary>
         public TempFile Beta { get; }
+
+        /// <summary>
+        /// An assembly with a reference to <see cref="Gamma"/>.
+        /// </summary>
         public TempFile Alpha { get; }
 
+        /// <summary>
+        /// An assembly with no references, assembly version 2.
+        /// </summary>
         public TempFile Delta2 { get; }
+
+        /// <summary>
+        /// An assembly with a reference to <see cref="Delta2"/>.
+        /// </summary>
         public TempFile Epsilon { get; }
 
+        /// <summary>
+        /// An assembly with no references, assembly version 2. The implementation however is different than
+        /// <see cref="Delta2"/> so we can test having two assemblies that look the same but aren't.
+        /// </summary>
         public TempFile Delta2B { get; }
+
+        /// <summary>
+        /// An assembly with no references, assembly version 3.
+        /// </summary>
         public TempFile Delta3 { get; }
 
         public TempFile UserSystemCollectionsImmutable { get; }

--- a/src/Tools/ExternalAccess/OmniSharp/Analyzers/OmnisharpAnalyzerLoaderFactory.cs
+++ b/src/Tools/ExternalAccess/OmniSharp/Analyzers/OmnisharpAnalyzerLoaderFactory.cs
@@ -1,0 +1,16 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.OmniSharp.Analyzers
+{
+    internal static class OmnisharpAnalyzerAssemblyLoaderFactory
+    {
+        public static IAnalyzerAssemblyLoader CreateShadowCopyAnalyzerAssemblyLoader(string? baseDirectory = null)
+        {
+            return new ShadowCopyAnalyzerAssemblyLoader(baseDirectory);
+        }
+    }
+}

--- a/src/VisualStudio/Core/Test.Next/Remote/SnapshotSerializationTests.cs
+++ b/src/VisualStudio/Core/Test.Next/Remote/SnapshotSerializationTests.cs
@@ -800,7 +800,7 @@ namespace Microsoft.CodeAnalysis.Remote.UnitTests
 
         private class MissingAnalyzerLoader : AnalyzerAssemblyLoader
         {
-            protected override Assembly LoadFromPathImpl(string fullPath)
+            protected override Assembly LoadFromPathUncheckedImpl(string fullPath)
                 => throw new FileNotFoundException(fullPath);
         }
 

